include/zephyr/drivers/input.h

```diff
@@ -0,0 +1,307 @@
+/*
+ * Copyright (c) 2022, tangchunhui@coros.com
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Public API for input devices.
+ * The scope of this API is report which input event was triggered
+ * and users can resolve input event by event type in their application.
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_INPUT_H_
+#define ZEPHYR_INCLUDE_DRIVERS_INPUT_H_
+
+#include <time.h>
+#include <stddef.h>
+#include <zephyr/types.h>
+#include <zephyr/device.h>
+
+
+#ifdef CONFIG_NEWLIB_LIBC
+
+#include <newlib.h>
+
+#ifdef __NEWLIB__
+#include <sys/_timeval.h>
+#else /* __NEWLIB__ */
+#include <sys/types.h>
+/* workaround for older Newlib 2.x, as it lacks sys/_timeval.h */
+struct timeval {
+	time_t tv_sec;
+	suseconds_t tv_usec;
+};
+#endif /* __NEWLIB__ */
+
+#else /* CONFIG_NEWLIB_LIBC */
+
+#ifdef CONFIG_ARCH_POSIX
+#include <bits/types/struct_timeval.h>
+#else
+#include <sys/_timeval.h>
+#endif
+
+#endif /* CONFIG_NEWLIB_LIBC */
+
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief INPUT APIs
+ * @defgroup input_interface Input Subsystem Driver APIs
+ * @ingroup io_interfaces
+ * @{
+ */
+
+/**
+ * @brief Input attribute types.
+ *
+ */
+enum input_attr_type {
+	/* Input driver hardware chipid, readonly. */
+	INPUT_ATTR_HARDWARE_CHIPID,
+	/* Input driver firmware version, readonly. */
+	INPUT_ATTR_FIRMWARE_VERSION,
+	/* Input device event read timeout */
+	INPUT_ATTR_EVENT_READ_TIMEOUT,
+};
+
+/**
+ * @brief Input attribute data union.
+ *
+ */
+union input_attr_data {
+	uint32_t chipid;
+	uint32_t version;
+	/* Waiting period to read event, or one of the special values K_NO_WAIT and K_FOREVER. */
+	k_timeout_t timeout;
+	/**
+	 * This is required to convert between custom data and attribute data.
+	 * Custom data size must be less than or equal to CONFIG_INPUT_ATTR_DATA_SIZE.
+	 */
+	uint32_t userdata[CONFIG_INPUT_ATTR_DATA_SIZE / sizeof(uint32_t)];
+};
+
+BUILD_ASSERT(sizeof(union input_attr_data) <= CONFIG_INPUT_ATTR_DATA_SIZE);
+
+/**
+ * @brief Representation of a input event.
+ *
+ */
+struct input_event {
+	/* The time of the event, the time is the system uptime */
+	struct timeval time;
+	/* event type (EV_SYN, EV_KEY, EV_ABS, etc) */
+	uint16_t type;
+	/* event code (Keyboard code, ABS code, etc) */
+	uint16_t code;
+	/* event value (Keyboard value, ABS position, etc) */
+	int32_t value;
+};
+
+#define input_event_size	sizeof(struct input_event)
+
+/**
+ * @brief Event type
+ *
+ */
+#define EV_SYN			0x00
+#define EV_KEY			0x01
+#define EV_REL			0x02
+#define EV_ABS			0x03
+
+/**
+ * @brief Synchronization events
+ *
+ */
+#define SYN_REPORT		0x00
+
+/**
+ * @brief Relative axes
+ *
+ */
+#define REL_X			0x00
+#define REL_Y			0x01
+#define REL_Z			0x02
+
+/**
+ * @brief Absolute axes
+ *
+ */
+#define ABS_X			0x00
+#define ABS_Y			0x01
+#define ABS_Z			0x02
+
+/**
+ * @brief Event Key value
+ *
+ */
+enum keyboard_value {
+	KEY_RELEASE,
+	KEY_PRESSED,
+	KEY_LONG_PRESSED,
```

Comment:
Why are these added. As soon as you have long press/release, why not support double tap? Triple? 

If the goal is a "lowest common denominator" API good enough for basic input to programs, this is probably fine.

I think maybe this is my disconnect, I come from developing featureful input devices. 

If the goal of this API is to focus on the *consumers* of input data, then we should clearly document that, since the current system doesn't seem to be flexible/powerful enough to handle the other.